drivers/serial/uart_b91.c

```diff
@@ -0,0 +1,571 @@
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	return (uart_b91_get_rx_bufcnt(uart) > 0) ? 1 : 0;
+}
+
+/* API implementation: irq_err_enable */
+static void uart_b91_irq_err_enable(const struct device *dev)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	uart->rxtimeout_h |= UART_ERR_IRQ_MASK;
+}
+
+/* API implementation: irq_err_disable*/
+static void uart_b91_irq_err_disable(const struct device *dev)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	uart->rxtimeout_h &= ~UART_ERR_IRQ_MASK;
+}
+
+/* API implementation: irq_is_pending */
+static int uart_b91_irq_is_pending(const struct device *dev)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	return ((uart->status & UART_IRQ_STATUS) != 0) ? 1 : 0;
+}
+
+/* API implementation: irq_update */
+static int uart_b91_irq_update(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* nothing to be done */
+	return 1;
+}
+
+/* API implementation: irq_callback_set */
+static void uart_b91_irq_callback_set(const struct device *dev,
+				      uart_irq_callback_user_data_t cb,
+				      void *cb_data)
+{
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	data->callback = cb;
+	data->cb_data = cb_data;
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api uart_b91_driver_api = {
+	.poll_in = uart_b91_poll_in,
+	.poll_out = uart_b91_poll_out,
+	.err_check = uart_b91_err_check,
+	.configure = uart_b91_configure,
+	.config_get = uart_b91_config_get,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_b91_fifo_fill,
+	.fifo_read = uart_b91_fifo_read,
+	.irq_tx_enable = uart_b91_irq_tx_enable,
+	.irq_tx_disable = uart_b91_irq_tx_disable,
+	.irq_tx_ready = uart_b91_irq_tx_ready,
+	.irq_tx_complete = uart_b91_irq_tx_complete,
+	.irq_rx_enable = uart_b91_irq_rx_enable,
+	.irq_rx_disable = uart_b91_irq_rx_disable,
+	.irq_rx_ready = uart_b91_irq_rx_ready,
+	.irq_err_enable = uart_b91_irq_err_enable,
+	.irq_err_disable = uart_b91_irq_err_disable,
+	.irq_is_pending = uart_b91_irq_is_pending,
+	.irq_update = uart_b91_irq_update,
+	.irq_callback_set = uart_b91_irq_callback_set,
+#endif
+};
+
+
+#define UART_B91_INIT(n)							    \
+										    \
+	static void uart_b91_irq_connect_##n(void);				    \
+										    \
+	static const struct uart_b91_config uart_b91_cfg_##n =			    \
+	{									    \
+		.uart_addr = DT_INST_REG_ADDR(n),				    \
+		.baud_rate = DT_INST_PROP(n, current_speed),			    \
+		.pirq_connect = uart_b91_irq_connect_##n			    \
+	};									    \
+	static struct uart_b91_data uart_b91_data_##n;				    \
+										    \
+	DEVICE_DT_INST_DEFINE(n, uart_b91_driver_init,				    \
+			      NULL,				    \
+			      &uart_b91_data_##n,				    \
+			      &uart_b91_cfg_##n,				    \
+			      PRE_KERNEL_1,					    \
+			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		    \
+			      (void *)&uart_b91_driver_api);			    \
+										    \
+	static void uart_b91_irq_connect_##n(void)				    \
```

Comment:
Looks like you are already familiar with the technique :)

Reply:
Yes :) 
It was not implemented for GPIO because I had issues with ``#if IS_INST_IRQ_EN(n)`` usage inside ``DT_NUM_INST_STATUS_OKAY(..)``. But now, looks like, I have provided good solution for GPIO.